### PR TITLE
Adds STS info for Operator

### LINKED ISSUE #792: Operator: SDKs can authenticate IAM-style to the Operator
Operator release v5.0.0 added [PR #1414](https://github.com/minio/operator/pull/1414) which allows for STS style uses for authenticating with the Operator.

We now support applications inside a kubernetes cluster to access MinIO with an IAM-style assume role mechanism.

1. Clients should read `AWS_WEB_IDENTITY_TOKEN_FILE` which points to the SA JWT token on the pod
2. client SDK should send the JWT over to the operator via custom endpoint which should be on the Operator
3. Operator should return an STS session if the service-account provided via the pod SA JWT has been assigned any policy

- [ ] Create a K8s-specific Developer doc for this functionality

More details are available at https://github.com/miniohq/engineering/issues/929.

### PR DESCRIPTION
**Pull request by djwfyi**
Creates a new page for STS tokens specifically for Operator.

Closes #792

Staged: http://192.241.195.202:9000/staging/k8s-sts/developers/sts-for-operator.html

### CODE REVIEW

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [inline source/developers/sts-for-operator.rst:27]:
```suggestion
An application can use an ``AssumeRoleWithWebIdentity`` call to send a request for temporary credentials to the MinIO Operator.
```

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [inline source/developers/sts-for-operator.rst:78]:
```suggestion
5. Use an SDK that supports the ``AssumeRoleWithWebIdentity`` like behavior to send a call from your application to the deployment
```

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [inline source/developers/sts-for-operator.rst:80]:
```suggestion
   Some SDKs that support ``AssumeRoleWithWebIdentity`` include:
```

**Review by pjuarezd — CHANGES_REQUESTED** [review]
Is possible to include the diagram?

https://github.com/minio/operator/blob/master/docs/images/sts-diagram.png

**Comment by djwfyi** [comment]
> Is possible to include the diagram?
> 
> https://github.com/minio/operator/blob/master/docs/images/sts-diagram.png

I've asked @oscarocastellanos to brandify it, then will add it when he's done.

File: source/index.rst
Comment (ravindk89) [inline source/index.rst:188]: Move this down to where line 203 is and it *should* work.

File: source/developers/sts-for-operator.rst
Comment (feorlen) [inline source/developers/sts-for-operator.rst:34]: Is there a "first reference" expansion for `JWT` nearby?

**Comment by oscarocastellanos** [comment]
Here's the diagram @djwfyi @pjuarezd [sts-diagram.zip](https://github.com/minio/docs/files/11385073/sts-diagram.zip)
sts-diagram
sts-diagram

File: source/developers/sts-for-operator.rst
Comment (feorlen) [inline source/developers/sts-for-operator.rst:61]: How would someone know what `the value for your MinIO deployment` should be? Would this be obvious to someone who knows Kubernetes? (Because that person is most assuredly not me.)

File: source/developers/sts-for-operator.rst
Comment (feorlen) [inline source/developers/sts-for-operator.rst:66]: Perhaps an article before `service account` here? Seems either `a` or `the` would work.

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:34]: good point @feorlen , JWT is a JSON web Token, but could expand the description of where it comes from, somenthing like this:


The STS API expects a JSON Web Token (JWT), it is a `Kubernetes service Account Token <https://kubernetes.io/docs/reference/access-authn-authz/authentication/#service-account-tokens>`.

When a pod have a Kubernetes service account associated to it, it will have the JWT  mounted in a well known location in the filesystem, that i  "/var/run/secrets/kubernetes.io/serviceaccount/token".

Take for example the following kubernetes manifest:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  namespace: sts-client
  name: my-service-account
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: default
spec:
  replicas: 3
  template:
    metadata:
    # ...
    spec:
      serviceAccountName: my-service-account
      containers:
      - name: nginx
        image: nginx:1.14.2

```
In the first yaml document of the manifest a service account `my-service-account` is created.
In the second yaml document a Deployment is created, and is associated to the service account `my-service-account` in the `spec.spec.serviceAccountName` field.

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:61]: For the mayority of the deployments, the operator is going to be on the `minio-operator` namespace by default.

On Openshift deployments coming from [OperatorHub](https://operatorhub.io/operator/minio-operator) the default namespace is `openshift-operators`.

Otherwise, users could use a different namespace with the `kubectl minio init --namespace other-namespace` command https://min.io/docs/minio/kubernetes/upstream/reference/kubectl-minio-plugin/kubectl-minio-init.html#kubectl.minio.init.-namespace.

File: source/developers/sts-for-operator.rst
Comment (feorlen) [inline source/developers/sts-for-operator.rst:17]: `any` could maybe be dropped here, it doesn't feel like it's adding much.

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:63]: Link to MinIO IAM policy here?

File: source/developers/sts-for-operator.rst
Comment (feorlen) [reply source/developers/sts-for-operator.rst:34]: @pjuarezd ooo nice extended description. A bit long to fit in this list, but maybe we can find somewhere useful. @djwfyi ? 

Thanks!

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:15]: Q: @pjuarezd what problem is this solving in K8s vs sending the API request to the tenant itself?

Also, does the `AssumeRoleWithWebIdentity` API remain the same? How does the application specify the tenant it wants to auth against?

**Review by ravindk89 — CHANGES_REQUESTED** [review]
Mostly clarifying questions.

File: source/developers/sts-for-operator.rst
Comment (feorlen) [inline source/developers/sts-for-operator.rst:47]: I keep wanting to read this as `The deployment`. I'll let you decide if adding the definite article works in this list.

Comment (feorlen) [reply source/developers/sts-for-operator.rst:47]: Also a tiny punctuation nit: period at the end of list items, or not?

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:15]: this is a excelent question a bit longer to answer, sorry for the long text.

`Operator STS API` solves some problems:

* Authenticate applications in kubernetes to access MinIO using a kubernetes native authentication mechanism: using kubernetes Service Accounts Tokens, the `Service Account` is a core concept of the Role-based Access Control (RBACK) Native kubernetes Authenticatio and authorization.

* Allows `Infrastructure as code` like configuration to access MinIO in kubernetes, by using  the Tenant CRD + PolicyBinding CRD.

* Allows a "credential-less" authentication in the client code, no need to create and specify IAM user to gain access to MinIO API, just instantiate the SDK client and get access to the 

We had often been asked to provide a way to get credentials for an application in a "credentialless" way, similar to how AWS IMDS allows to get the STS Token of the instance profile assigned to the EC2 instance. [AWS doc example here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials)

We also have been asked to provide an mechanism to authenticate using kubernetes Service accounts similar to [AWS IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html), Operator STS does it.

Comment (ravindk89) [reply source/developers/sts-for-operator.rst:15]: Thanks @pjuarezd - so that comes back to how to use the API itself. From what I'm reading in a previous comment, the application would need to have access to the JWT from Kubernetes itself - so that JWT has to be passed to the applications somehow and then maintained in case it ever changes.

that still leaves a question of how the Operator knows which Tenant to generate credentials for. Is it purely by the service account token being namespaced already?

Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:15]: yeah, regarding that second question

> Also, does the AssumeRoleWithWebIdentity API remain the same? How does the application specify the tenant it wants to auth against?

The request is not the same that doing it directly to the tenant, as the tenant knows nothing about kuberneters RBAC, Operator will do that authentication, and then request an STS token to the tenant using root credentials, somenthing like this:

Authentication with service Accoun

Comment (ravindk89) [reply source/developers/sts-for-operator.rst:15]: That makes sense. I guess here then the one requirement is we have to assume the K8s admin has a way of passing that SA JWT to the application and keeping it up-to-date, otherwise things crap out. @djwfyi I would suggest mentioning something to that effect if it's not there already and I just dumb brained past it.

Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:15]: The bridge to know which tenant the sdk App is getting credentials for is the PolicyBinding Custom resource.

1) The client specify the namespace of the tenant in the STS endpoint, ie using the AWS SDK code would look like this

```python
stsUrl = urlparse(f"https://sts.minio.operator.svc.cluster.local:4223/tenant-1")
sts = boto3.client('sts', endpoint_url=stsUrl.geturl(), verify=False)
assumed_role_object = sts.assume_role_with_web_identity(
    RoleArn="arn:aws:iam::111111111:dummyroot",
    RoleSessionName="session 1",
    WebIdentityToken=sa_jwt
)

s3_client = boto3.resource('s3',
    aws_access_key_id=credentials['AccessKeyId'],
    aws_secret_access_key=credentials['SecretAccessKey'],
    aws_session_token=credentials['SessionToken'],
    endpoint_url=tenantUrl.geturl(), verify=False)
```
2) a `PolicyBinding` object have to exist in the tenant namespace, and specify service account name of the JWT, and a list of IAM policies to authorize.


```yaml
apiVersion: sts.min.io/v1alpha1
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: tenant-1
spec:
  application:
    namespace: application-namespace
    serviceaccount: aplication-service-account
  policies:
    - bucket1-rw
    - bucket2-read
```


Because the PolicyBinding is a Custom resource and can be specified as yaml, this opens the door to Infrastructure as code authentication, managed by the kubernetes manager. I am thinking of ArgoCD maybe?

Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:15]: > That makes sense. I guess here then the one requirement is we have to assume the K8s admin has a way of passing that SA JWT to the application and keeping it up-to-date, otherwise things crap out. @djwfyi I would suggest mentioning something to that effect if it's not there already and I just dumb brained past it.

we not necesarelly need to assume the k8s admin has a way of passing the SA JWT, because the application running in a pod in kubernetes is the one that will have to pass the JWT to the SDK's, for that, the pods always have the JWT   mounted in the filesystem  in the path `/var/run/secrets/kubernetes.io/serviceaccount/token`

I did a bit of explanation of where the JWT comes from in this comment https://github.com/minio/docs/pull/841#discussion_r1183996140

Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:15]: STS are temporary credentials, that expire and need to be recreated, SDK's currently have a verification flow, if the STS token is expired, then sdk library "should" requests  a new STS token before the next API call, but that is up to how the SDK client is implemented.

SA JWT also has expiration, yes that token expiration is a whole topic, there is long-lived tokens and short-lived tokens, automatic mounted tokens, manually created, etc. But we could just limit ourselfs to require a `valid, non expired Service account Token` regardless if is the one mounted in the pod, or another requested using K8s API by the SDK application.

This medium article might do a better explaining to Service account JWT that me https://medium.com/@th3b3ginn3r/understanding-service-accounts-in-kubernetes-e9d2abe19df8

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:34]: Thanks for this, @pjuarezd.
I expanded the first instance of JWT, then included portions of this description in several places.
For example, part of it in the last step of the procedure where the SDK call happens. And more of it in the Service Account description under Example Resources.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:61]: Added some instructions here for namespace and deployment values.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:66]: Yeah, and a period at the end. 😬

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:17]: Agreed

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:15]: This is all great info. Added applicable changes throughout the PR.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:47]: Me and listing ending punctuation have a long and fraught history that lacks anything resembling consistency.
Fixed both here.

**Review by feorlen — APPROVED** [review]
👍🏻

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [inline source/developers/sts-for-operator.rst:103]: Maybe is useful mention that if you define the environment variable
`AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/kubernetes.io/serviceaccount/token` then you don't need to read the JWT token file manually and pass it to the SDK client as parameter.

For example AWS SDK describes it here:
https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html#envvars-list-AWS_WEB_IDENTITY_TOKEN_FILE

On the minio Go SDK we hare an example in the tests

https://github.com/minio/minio-go/blob/8c4c0768b1b1d29b52457d8e048144a8e2901009/pkg/credentials/iam_aws_test.go#L353-L355

not sure how much detailed or extended should be in the docs that mention, but for sure we can have code examples of this on the Operator repo, and just mention "oh you can do AWS_WEB_IDENTITY_TOKEN_FILE too" or somenthing like that.

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [inline source/developers/sts-for-operator.rst:124]: Rather than save it as a file in the pod file system, k8s mounts it, like this for example

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: nginx
  name: nginx-deployment-7fb96c846b-mvjhs
  namespace: default
spec:
  containers:
  - image: nginx:1.14.2
    name: nginx
...
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-vsvfz
      readOnly: true
...
  volumes:
  - name: kube-api-access-vsvfz
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
...
```

```suggestion
When linked to a pod, such as through a deployment's ``.spec.spec.serviceAccountName`` field, Kubernetes mounts a :abbr:`JWT (JSON Web Token)` for the service account in a well-known location, such as ``/var/run/secrets/kubernetes.io/serviceaccount/token``.
```

**Review by pjuarezd — COMMENTED** [review]
only suggestions, no hard requirements to change anything, wil let you do the switfness of wording it better than me @djwfyi

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:38]: I think here is where we should state that the API call must include the ServiceAccount JWT, which applications deployed in the same cluster can access through the URL @pjuarezd mentioned.

Without that key element, users wont fully understand what to present.

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:49]: Or maybe this is where we need to clarify - from what I'm gathering from Pedro's comments, it's the same AssumeRoleWithWebToken API, but the JWT is the ServiceAccount token for the target tenant.

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:84]:
```suggestion
2. Ensure an appropriate :ref:`policy <minio-policy>` exists on the MinIO Tenant for the application to use.
   This procedure includes associating that policy to the ServiceAccount through a PolicyBinding object.
```

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:86]:
```suggestion
3. Create YAML resources for the Service Account and Policy Binding: 
```

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:93]: :thought_balloon: @pjuarezd more follow-ups - if I have `minio-tenant-1` namespace and `application-1` namespace, which namespace gets the SA and policy binding?

From the examples below it looks like we apply it to the client/application namespace. Just want to confirm, because if so @djwfyi I suggest we state that here. Even if we have examples below, it's easy to miss. Should also state that in the examples, as by default w/o namespace those will end up in the default.

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:103]:
```suggestion
   This is typically in a well known location on pods in the Kubernetes cluster, such as ``/var/run/secrets/kubernetes.io/serviceaccount/token``.
```

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:105]:
```suggestion
   The following MinIO SDKs support ``AssumeRoleRoleWithWebIdentity`` :
```

**Review by ravindk89 — CHANGES_REQUESTED** [review]
Strictly speaking these might not be necessary to go - you can take this in a separate PR if we want to get this out sooner. But I think it would help with removing a few more stumbling blocks.

File: source/developers/sts-for-operator.rst
Comment (pjuarezd) [reply source/developers/sts-for-operator.rst:93]: Good observation, the PolicyBinding resource has to be created in the tenant namespace, that is `minio-tenant-1`.
In Daryl's example it will be like this.

```yaml
apiVersion: sts.min.io/v1alpha1
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: minio-tenant-1
spec:
  application:
    namespace: sts-client
    serviceaccount: stsclient-sa
  policies:
    - test-bucket-rw
```

The `application` we are granting access to is stated in the combination of `.spec.application.serviceaccount` and `.spec.application.namespace`, service accounts are namespace bounded, only one service account with certain name can exist in a namespace.

For instance, in the above Daryl's example, the service account with name `sts-client-sa` has to exist in the namespace `sts-client`,  and the manifest would be like this:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  namespace: sts-client
  name: stsclient-sa
```

Now, there is no restriction of in which namespace the service account have to be in, it could be even in the same tenant namespace, it doesn't matter. The only that matters is that the PolicyBinding states correctly the name of the service account, and the namespace where it lives at.

Comment (ravindk89) [reply source/developers/sts-for-operator.rst:93]: Makes sense - I'm not particularly familiar with how much namespace isolation affects these things. I assume that as long as the application has the correct kube permissions (however those might work) to access the SA, the rest just sort of happens on its own.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:91]: Rephrased this line to include that info and also added an additional sentence to the Policy Binding section down the page.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:103]: Added some info too this effect in this step of the procedure.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:33]: Tweaked slightly, but added.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:49]: I mentioned this earlier on the page now, so leaving it out here for now.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:84]: I'll add something to this effect, but I find it kind of awkward because the very next step does this.

File: source/developers/sts-for-operator.rst
Comment (djwfyi) [reply source/developers/sts-for-operator.rst:93]: Corrected my description of the YAML for the policy binding, which was wrong.
Also added inline comments to the YAML to better clarify what namespace to use.

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:40]:
```suggestion
An application can use an ``AssumeRoleWithWebIdentity`` call including a Kubernetes Service Account's :abbr:`JWT (JSON Web Token)` to send a request for temporary credentials to the MinIO Operator.
```

Maybe link to the service account docs here

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:41]: From what Pedro said, it's more that

"Kubernetes mounts Service Accounts to pods specifying that SA using <whatever that field was>.
The Pod can access those SAs at a well known location such as <URL>"

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:53]:
```suggestion
1. An application sends an AssumeRoleWithWebidentity API request to the MinIO Operator containing the tenant namespace and a service account to use.
```

Link to the API docs.

File: source/developers/sts-for-operator.rst
Comment (ravindk89) [inline source/developers/sts-for-operator.rst:109]: For pods which specify the service account using <that field>, Kubernetes typically mounts that JWT at a well known location such as <thing>

**Review by ravindk89 — APPROVED** [review]
Mostly nits, but LGTM once you take/leave them.

**Review by pjuarezd — APPROVED** [review]
LGTM!

**Pull request merged**
